serialization/pickle.ksy

```diff
@@ -0,0 +1,443 @@
+meta:
+  id: pickle
+  title: Python pickle serialiazation format (protocol 3)
+  application: Python
+  file-extension:
+    - pickle
+    - pkl
+    #- pkl0
+    #- pkl1
+    #- pkl2
+  license: CC0-1.0
+  endian: le
+doc: |
+  Python Pickle format serializes Python objects to a byte stream, as a sequence
+  of operations to run on the Pickle Virtual Machine.
+
+  The format is mostly implementation defined, there is no formal specification.
+  Pickle data types are closely coupled to the Python object model.
+  Python singletons, and most builtin types (e.g. `None`, `int`,`dict`, `list`)
+  are serialised using dedicated Pickle opcodes.
+  Other builtin types, and all classes  (e.g. `set`, `datetime.datetime`) are
+  serialised by encoding the name of a constructor callable.
+  They are deserialised by importing that constructor, and calling it.
+doc-ref:  https://github.com/python/cpython/blob/3.3/Lib/pickletools.py
+seq:
+  # TODO is there a way to declare PROTO is optional, but only valid at position 0?
+  - id: ops
+    type: op
+    repeat: eos
+  # TODO is there a way to declare a trailing STOP is required?
```

Comment:
[`contents`](https://doc.kaitai.io/user_guide.html#_checking_for_magic_signatures)

Reply:
This doesn't work in combination with `repeat: eos` though, right? (The array will always consume all of the data, so the `contents` field will never have any data left and will always fail.)

If `STOP` can only appear at the end of the data, you could use `repeat: until` to terminate the array after the first `STOP`.

Reply:
`_io` magic is to the rescue. One can determine the stream size and create a substream containing all but the last bytes. So the last bytes are to be used for `contents` and the rest is occupied by a substream containing `repeat: eos`